runtime/ops/tls.rs

```diff
@@ -70,6 +88,457 @@ impl StoresClientSessions for ClientSessionMemoryCache {
+  }
+
+  fn new_server_side(tcp: TcpStream, tls_config: &Arc<ServerConfig>) -> Self {
+    let tls = TlsSession::Server(ServerSession::new(tls_config));
+    Self::new(tcp, tls)
+  }
+
+  fn into_split(self) -> (ReadHalf, WriteHalf) {
+    let shared = Shared::new(self);
+    let rd = ReadHalf {
+      shared: shared.clone(),
+    };
+    let wr = WriteHalf { shared };
+    (rd, wr)
+  }
+
+  fn poll_io(
+    &mut self,
+    cx: &mut Context<'_>,
+    flow: Flow,
+  ) -> Poll<io::Result<()>> {
+    // Initially, assume that the TCP stream is both readable and writable. If
+    // this turns out not to be the case, `read_tls` or `write_tls()` will fail
+    // with `WouldBlock`; only then do we arrange to be asynchronously notified
+    // of read/write.
+    let mut tcp_read_ready = true;
+    let mut tcp_write_ready = true;
+
+    loop {
+      while tcp_write_ready && self.tls.wants_write() {
+        let mut wrapper = ImplementWriteTrait(&mut self.tcp);
+        match self.tls.write_tls(&mut wrapper) {
+          Ok(0) => unreachable!(),
+          Ok(_) => {}
+          Err(err) if err.kind() == ErrorKind::WouldBlock => {
+            match self.tcp.poll_write_ready(cx) {
+              Poll::Ready(result) => result?,
+              Poll::Pending => tcp_write_ready = false,
+            }
+          }
+          Err(err) => return Poll::Ready(Err(err)),
+        }
+      }
+
+      while tcp_read_ready
+        && self.tls.wants_read()
+        && self.rd_shut != Shut::TcpShut
+      {
+        let mut wrapper = ImplementReadTrait(&mut self.tcp);
+        match self.tls.read_tls(&mut wrapper) {
+          Ok(0) if self.tls.is_handshaking() => {
+            let err =
+              io::Error::new(ErrorKind::UnexpectedEof, "tls handshake eof");
+            return Poll::Ready(Err(err));
+          }
+          Ok(0) => self.rd_shut = Shut::TcpShut,
+          Ok(_) if matches!(self.rd_shut, Shut::TlsShut | Shut::TcpShut) => {
+            // Just like TCP, when data arrives after the connection has been
+            // closed it means that the connection wasn't closed in a clean way.
+            let err = io::Error::from(ErrorKind::ConnectionReset);
+            return Poll::Ready(Err(err));
+          }
+          Ok(_) => match self.tls.process_new_packets() {
+            // Do a zero-length plaintext read so we can detect the arrival of
+            // 'CloseNotify' messages, even if only the write half is active.
+            // Actually reading data from the socket is done in `poll_read()`.
+            Ok(()) => match self.tls.read(&mut []) {
+              Ok(0) => {}
+              Ok(_) => unreachable!(),
+              // Rustls `read()` returns `ConnectionAborted` when it receives a
+              // 'CloseNotify' alert; this indicates that the remote end of the
+              // connection has initiated a graceful shutdown at the TLS level.
+              Err(err) if err.kind() == ErrorKind::ConnectionAborted => {
```

Comment:
We cannot depend on `ConnectionAborted` to determine eof, because tls stream may not receive close notify.

Reply:
13 lines up graceful close of the underlying TCP connection gets handled. https://github.com/denoland/deno/blob/6cd7b48744db8718b243dfff8afb676a039d916e/runtime/ops/tls.rs#L275

Reply:
Tcp shutdown does not mean Tls shutdown, I mentioned this in another comment.

Reply:
Well, I reconsidered it. If `wants_read` guarantees that only one packet is processed in rustls, then it is fine.